### PR TITLE
RUMM-506 Prevent malforming payload due to I/O exception 

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

🐛  Some payloads send from SDK were malformed.

The reason was not performing writes atomically:
```swift
// pseudo-code

let separatorData = ",".utf8Data
try writer.write(separatorData)
try writer.write(logData)
```
In rare cases, when the `I/O` exception was thrown for the second write, it was leaving the file's content malformed.

### How?

The fix is to perform atomic writes:
```swift
// pseudo-code

let logWithSeparatorData = ",".utf8Data + logData
try writer.write(logWithSeparatorData)
```

The red unit test for payload malformation was added, then it was made green with a fix.

### Later improvements

The `File` interface is not resistant for this sort of bugs, as it doesn't perform transactional write even if it looks like:
```swift
// FileTests.swift

try file.append { write in
    try write(Data([0x42, 0x42, 0x42, 0x42, 0x42])) // 5 bytes
    try write(Data([0x41, 0x41, 0x41, 0x41, 0x41])) // 5 bytes
}
```

As this requires additional refactoring, to not impact the hotfix scope, I will open a separate PR changing the `WritableFile` interface from:
```swift
func append(transaction: ((Data) throws -> Void) throws -> Void) throws
```
to simply:
```swift
func append(data Data) throws
```

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Sources/Datadog/Core/Persistence/FileWriter.swift [hunk 1]
```diff
@@ -42,9 +42,9 @@ internal final class FileWriter {
                     try write(data)
                 }
             } else {
+                let atomicData = commaSeparatorData + data
                 try file.append { write in
-                    try write(commaSeparatorData)
-                    try write(data)
+                    try write(atomicData)
                 }
```
Comment (buranmert) [inline Sources/Datadog/Core/Persistence/FileWriter.swift:48]: we can also reduce `try write()` calls
```swift
let rawData = try jsonEncoder.encode(value)
let data: Data = try file.size() == 0 ? (rawData) : (commaSeparator + data)
try file.append { write in try write(data) }
```
wdyt?

Comment (ncreated) [reply Sources/Datadog/Core/Persistence/FileWriter.swift:48]: Good idea 👍. However, to keep the hotfix scope as small as possible, let's do it in a separate PR to `master` (within the "Later improvements" I list in the PR description).

**Comment by buranmert** [comment]
@ncreated i just noticed that in some cases the log has `\u0000` instead of extra `,`.
i guess that's because the second `try write(data)` throws and leaves `\u0000` there, can this be true? 
if so, how can we make sure that we don't leave corrupt chars in the file?

**Comment by buranmert** [comment]
note: we still can't figure out why `\u0000` appears, we couldn't reproduce it and it seems very rare (possibly a few in a million logs)

we will ship this hotfix version and keep monitoring the issue

**Comment by ncreated** [comment]
I added the test to prevent also malformations in the middle of the file: https://github.com/DataDog/dd-sdk-ios/pull/133/commits/792c74cd886372c3d002305a2eb970c218bbac34 This is to additionally check if the `0x00` byte is not inserted.

**Comment by buranmert** [comment]
i suspect that when `FileHandler.write(data)` raises a certain kind of exception, it somehow changes `offsetInFile` without adding non-null bytes.

can we add something like this?
```swift
func append(transaction: ((Data) throws -> Void) throws -> Void) throws {
  let fileHandle = try FileHandle(forWritingTo: url)
  let initialOffset = fileHandle.offsetInFile
  defer {
    let finalOffset = fileHandle.offsetInFile
    if errorThrown && (initialOffset != finalOffset) {
      userLogger.error("Houston, we have a problem...")
    }
    fileHandle.closeFile()
  }
  ...
}       
```

**Comment by ncreated** [comment]
> can we add something like this?

@buranmert even if it works, this does neither fix the problem nor empower us with reporting. I think it's better to iterate on this issue and first see if we'll be receiving malformed logs after applying this hotfix. WDYT?

**Comment by buranmert** [comment]
yes, fixing null char issue doesn't seem trivial; let's go with this hotfix and keep monitoring 👍

**Pull request merged**